doc/decisions/internal_cache.md

```diff
@@ -2,26 +2,98 @@
-no keys might be returned, because it is up-to-date.
+`kdbGet` might return more or less keys than requested.
+This was found confusing several times.
 
-When doing kdbSet() a deep duplication is needed.
+### More Keys
 
-Idea: keep a duplicated keyset internally. Return (non-deep?)
-duplications exactly of the parts as requested.
+Even if calling `kdbGet` with a parent key below a mountpoint, `kdbGet` will nevertheless return all keys of the mountpoint.
+Pseudo code example, assuming there is a mountpoint at `/mountpoint` and a key `/mountpoint/other`:
+
+```
+kdbGet (kdb, ks, keyNew("/mountpoint/below"));
+assert (ksLookup (ks, "/mountpoint/other") == NULL);
+```
+
+If was found unexpected that this assert will fail.
+
+### Less Keys
+
+When doing a second `kdbGet` with a new keyset no keys might be returned, because kdb internally is up-to-date.
+Pseudo code example, assuming there is a key `/somewhere/key`:
+
+```
+kdbGet (kdb, ks1, keyNew("/somewhere"));
+assert (ksLookup (ks1, "/somewhere/key") != NULL);
+kdbGet (kdb, ks2, keyNew("/somewhere"));
+assert (ksLookup (ks2, "/somewhere/key") != NULL);
+```
+
+If was found unexpected that the second assert will fail.
 
 ## Constraints
 
+- memory consumption must be low for `kdbGet`, see [4. Goal: Performance](/doc/GOALS.md), in particular, deep duplication is too expensive
+
 ## Assumptions
 
 ## Considered Alternatives
 
-- no cache (current situation)
-- flat cache with COW
-- deep duplicated cache
-- Use global plugin to implement internal cache.
-- Use global mmap plugin and make kdbGet() to always retrieve everything.
+### Keep Current Situation
+
+Improve documentation to make people more aware of these two problems:
+
+- add a tutorial about `kdbGet` semantics
+- add full examples how to correctly work with `kdbGet`
+
+### Cachefilter Plugin
+
+Naively one would simply cache the whole keyset and use `ksBelow` to always get the keyset.
+
+This idea was implemented and later on discarded: a3d95f07160d792fdd0f169d8543138c32a2f580
+
+The main problems are:
+
+- very high memory consumption (duplication of KeySets)
+- problems specific to [hooks](hooks.md), see [#1072](https://issues.libelektra.org/1072)
+
+### MMAP Cache with parent key
+
+We make the mmap cache non-optional so that we always have a keyset of configuration data internally.
+From this keyset, we use `ksBelow` to return the correct keyset.
+
+**Cons:**
+
+- invalidation of OPMPHM
+
+### MMAP Cache without parent key
+
+We make the mmap cache non-optional and only use a single cache, caching everything.
+We remove the parent key of `kdbGet` and `kdbSet` and always return the keyset of the whole KDB.
+
+### In-Memory COW Cache
+
+We keep a duplicated keyset in-memory and tag the keys as copy-on-write (COW).
+From this keyset, we use `ksBelow` to return the correct keyset.
+If the user tries to change the keys, the value or metadata gets duplicated, so that the original keyset is not changed.
```

Comment:
I guess the metadata question depends entirely on whether KeySets also are COW or not. If they are, your proposed approach will work fine.

If not, we need to `ksDup` and this will again waste memory.

Reply:
> I guess the metadata question depends entirely on whether KeySets also are COW or not.

Yup, either

- there is a normal `ksDup` in `keyCopy (..., ELEKTRA_CP_COW)` or `keyMeta` (when the keys is `KEY_FLAG_COW`),
- OR `keyCopy (..., ELEKTRA_CP_COW)` sets up `key->meta` to also be COW similar to what I describe above.

Whether or not we support COW for KeySets outside of the special case for metadata is a different question. For the `kdbGet` use case we don't need it.

Reply:
I clarified a bit fb272937de8dc760daacd21d7427dd4b46debb8d but I think the discussion is way too detailed for the current draft.